### PR TITLE
feat(PT-3253): allowance event

### PR DESCRIPTION
**Pull request by dmitry-saraykin**

### CODE REVIEW

File: erc20/core/src/main/kotlin/com/rarible/protocol/erc20/core/service/reduce/Erc20EventChainUpdateService.kt [hunk 1]
```diff
@@ -24,17 +26,21 @@ class Erc20EventChainUpdateService(
...
             val balanceId = BalanceId.parseId(entityId)
-            val lastEvent = erc20EventConverter.convert(
-                events.last().record.asEthereumLogRecord(),
-                events.last().eventTimeMarks.addIndexerIn()
-            )
-            val balance = erc20BalanceService.onChainUpdate(balanceId, lastEvent)
+            val erc20Events = events.map {
+                erc20EventConverter.convert(it.record.asEthereumLogRecord(), it.eventTimeMarks)
+            }
+            val approvalEvent = erc20Events.lastOrNull { it?.event is Erc20Event.Erc20TokenApprovalEvent }
```
Comment (mystdeim) [inline erc20/core/src/main/kotlin/com/rarible/protocol/erc20/core/service/reduce/Erc20EventChainUpdateService.kt:32]: I think it should be in descriptor not in reducer

File: erc20/core/src/main/kotlin/com/rarible/protocol/erc20/core/repository/Erc20AllowanceRepository.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.data.mongodb.core.ReactiveMongoTemplate
+import org.springframework.data.mongodb.core.findById
+import org.springframework.data.mongodb.core.query.Query
+import org.springframework.data.mongodb.core.query.isEqualTo
+import org.springframework.data.mongodb.core.remove
+import org.springframework.stereotype.Component
+import scalether.domain.Address
+
+@Component
+@CaptureSpan(type = "db", subtype = "erc20-allowance")
```
Comment (yakischik) [inline erc20/core/src/main/kotlin/com/rarible/protocol/erc20/core/repository/Erc20AllowanceRepository.kt:17]: remove

File: erc20/core/src/main/kotlin/com/rarible/protocol/erc20/core/service/Erc20BalanceService.kt [hunk 3]
```diff
@@ -47,10 +49,14 @@ class Erc20BalanceService(
...
         return optimisticLock {
-            val erc20Balance = get(contract = balanceId.token, owner = balanceId.owner) ?: run {
-                logger.error("Can't get $balanceId from db")
-                return@optimisticLock  null
-            }
+            val erc20Balance = get(contract = balanceId.token, owner = balanceId.owner) ?: Erc20Balance(
+                owner = balanceId.owner,
+                token = balanceId.token,
+                balance = EthUInt256(BigInteger.ZERO),
+                createdAt = Instant.now(),
```
Comment (yakischik) [inline erc20/core/src/main/kotlin/com/rarible/protocol/erc20/core/service/Erc20BalanceService.kt:56]: nowMillis()

File: erc20/core/src/main/kotlin/com/rarible/protocol/erc20/core/service/reduce/Erc20EventChainUpdateService.kt [hunk 4]
```diff
@@ -24,17 +26,21 @@ class Erc20EventChainUpdateService(
     private suspend fun handleBalanceEvent(entityId: String, events: List<LogRecordEvent>) {
         try {
             val balanceId = BalanceId.parseId(entityId)
-            val lastEvent = erc20EventConverter.convert(
-                events.last().record.asEthereumLogRecord(),
-                events.last().eventTimeMarks.addIndexerIn()
-            )
-            val balance = erc20BalanceService.onChainUpdate(balanceId, lastEvent)
+            val erc20Events = events.map {
+                erc20EventConverter.convert(it.record.asEthereumLogRecord(), it.eventTimeMarks)
```
Comment (yakischik) [inline erc20/core/src/main/kotlin/com/rarible/protocol/erc20/core/service/reduce/Erc20EventChainUpdateService.kt:30]: addIndexerIn()

File: erc20/core/src/main/kotlin/com/rarible/protocol/erc20/core/service/Erc20AllowanceService.kt [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            ?.let { EthUInt256(it) }
+    }
+
+    suspend fun onChainUpdate(balanceId: BalanceId, event: Erc20MarkedEvent?) {
+        optimisticLock {
+            val erc20Allowance = get(balanceId) ?: Erc20Allowance(
+                token = balanceId.token,
+                owner = balanceId.owner,
+                allowance = EthUInt256(BigInteger.ZERO),
+                createdAt = Instant.now(),
```
Comment (yakischik) [inline erc20/core/src/main/kotlin/com/rarible/protocol/erc20/core/service/Erc20AllowanceService.kt:67]: nowMillis()

File: erc20/core/src/main/kotlin/com/rarible/protocol/erc20/core/service/Erc20AllowanceService.kt [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    override suspend fun get(id: BalanceId): Erc20Allowance? =
+        if (saveToDb) {
+            erc20AllowanceRepository.get(id)
+        } else {
+            Erc20Allowance(
+                owner = id.owner,
+                token = id.token,
+                allowance = getBlockchainAllowance(id) ?: EthUInt256(BigInteger.ZERO),
+                createdAt = Instant.now(),
```
Comment (yakischik) [inline erc20/core/src/main/kotlin/com/rarible/protocol/erc20/core/service/Erc20AllowanceService.kt:39]: nowMillis()

**Review by yakischik — APPROVED** [review]

**Review by yakischik — APPROVED** [review]

**Pull request merged**